enhancements/20220125-concurrency-control-on-host-disk-operations.md

```diff
@@ -0,0 +1,271 @@
+# Concurrency Control on Host Disk Operations
+
+## Summary
+
+Some operations, such as formatting, might take a while for operating system to
+finish. It causes problems if the controller of host disk manager re-enters
+these criticial sections. To prevent re-entrance, additional disk management of
+Harvester needs a robust concurrency control on host disk operations.
+
+### Related Issues
+
+- Original bug report: https://github.com/harvester/harvester/issues/1718
+- Previous PR: https://github.com/harvester/node-disk-manager/pull/23
+
+## Motivation
+
+### Goals
+
+- Prevent criticial disk operations from re-entrances.
+- Control maximum count of concurrent disk operations within a time period.
+- Refactor the logic of blockdevice reconcilation to a more deterministic,
+  readable, and maintainable state.
+
+### Non-goals
+
+- Auto-recovering from disk operation failures is not in the scope. Node disk
+  manager does not know how to deal with this kind of failures (might due to 
+  hardware or OS failures). User should come in and tackle the issue manually
+  before re-triggering any futher operation.
+
+## Proposal
+
+This proposal is originally for tackling [issue #1718]. During the investigation
+of that issue, we found several potential flaws in the current implementation:
+
+1. After introduced [feature auto-provision], user is able to format and
+   provision a bunch of raw block device at once. And that leads to several 
+   race conditions between each format or provision operation, which might fail
+   eventually and make disks result in a wrong state.
+2. Operations like formatting a disk may take time. When `OnBlockDeviceChange`
+   has been called the first time and the disk started formatting, and later it
+   was called the second time but the former formatting has not yet finished,
```

Comment:
There are two possibilities here:
1. The first format is in progress then the second one retried, which should be 100% prevented from the controller side.
2. The first format failed or succeed, then the second time the object was requeued. We only need to check if the device is formatted to know it's status. If it's not formated when we're handling the disk, then we format it. We won't format the same disk twice at the same time due to controller limits the access to the object.

One very common issue would be, the format operation itself succeed, but the update to the object failed due to update conflict, so the CR will be requeued. Even in this case, we don't need to redo the format in this case if we detect the device has already been formatted.